### ISSUE TITLE
pnnx导出gn出错

### ISSUE THREAD

**Issue by XdpAreKid** [Tencent/ncnn#4519]
## error log | 日志或报错信息 | ログ
pnnx rest.pt "inputshape=[1,3,512,512]" moduleop=gfpgan.archs.restoreformer_arch.VectorQuantizerEmbedding
pnnxparam = rest.pnnx.param
pnnxbin = rest.pnnx.bin
pnnxpy = rest_pnnx.py
pnnxonnx = rest.pnnx.onnx
ncnnparam = rest.ncnn.param
ncnnbin = rest.ncnn.bin
ncnnpy = rest_ncnn.py
fp16 = 1
optlevel = 2
device = cpu
inputshape = [1,3,512,512]f32
inputshape2 = 
customop = 
moduleop = gfpgan.archs.restoreformer_arch.VectorQuantizerEmbedding
############# pass_level0
inline module = gfpgan.archs.restoreformer_arch.Downsample
inline module = gfpgan.archs.restoreformer_arch.MultiHeadAttnBlock
inline module = gfpgan.archs.restoreformer_arch.MultiHeadAttnBlockTransformer
inline module = gfpgan.archs.restoreformer_arch.MultiHeadDecoderTransformer
inline module = gfpgan.archs.restoreformer_arch.MultiHeadEncoder
inline module = gfpgan.archs.restoreformer_arch.ResnetBlock
inline module = gfpgan.archs.restoreformer_arch.Upsample
inline module = gfpgan.archs.restoreformer_arch.VectorQuantizer
inline module = gfpgan.archs.restoreformer_arch.Downsample
inline module = gfpgan.archs.restoreformer_arch.MultiHeadAttnBlock
inline module = gfpgan.archs.restoreformer_arch.MultiHeadAttnBlockTransformer
inline module = gfpgan.archs.restoreformer_arch.MultiHeadDecoderTransformer
inline module = gfpgan.archs.restoreformer_arch.MultiHeadEncoder
inline module = gfpgan.archs.restoreformer_arch.ResnetBlock
inline module = gfpgan.archs.restoreformer_arch.Upsample
inline module = gfpgan.archs.restoreformer_arch.VectorQuantizer

----------------

############# pass_level1
no attribute value
no attribute value
no attribute value
unknown Parameter value kind prim::Constant of TensorType, t.dim = 2
no attribute value
no attribute value
unknown Parameter value kind prim::Constant
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
no attribute value
############# pass_ncnn
ignore nn.GroupNorm decoder.mid.attn_1.norm2 param affine=True
ignore nn.GroupNorm decoder.mid.attn_1.norm2 param eps=1.000000e-06
ignore nn.GroupNorm decoder.mid.attn_1.norm2 param num_channels=512
ignore nn.GroupNorm decoder.mid.attn_1.norm2 param num_groups=32
ignore nn.GroupNorm decoder.up.5.attn.0.norm2 param affine=True
ignore nn.GroupNorm decoder.up.5.attn.0.norm2 param eps=1.000000e-06
ignore nn.GroupNorm decoder.up.5.attn.0.norm2 param num_channels=512
ignore nn.GroupNorm decoder.up.5.attn.0.norm2 param num_groups=32
ignore nn.GroupNorm decoder.up.5.attn.1.norm2 param affine=True
ignore nn.GroupNorm decoder.up.5.attn.1.norm2 param eps=1.000000e-06
ignore nn.GroupNorm decoder.up.5.attn.1.norm2 param num_channels=512
ignore nn.GroupNorm decoder.up.5.attn.1.norm2 param num_groups=32
ignore nn.GroupNorm decoder.up.5.attn.2.norm2 param affine=True
ignore nn.GroupNorm decoder.up.5.attn.2.norm2 param eps=1.000000e-06
ignore nn.GroupNorm decoder.up.5.attn.2.norm2 param num_channels=512
ignore nn.GroupNorm decoder.up.5.attn.2.norm2 param num_groups=32

## model | 模型 | モデル
1. original model

[模型地址](https://drive.google.com/file/d/1bjWJrd9tVl6xNhzv9AdxWGmal8tGQTlX/view?usp=share_link)

**Comment by nihui** [comment]
a minimal test

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class Normalize(nn.Module):
    def __init__(self):
        super().__init__()

        self.norm = nn.GroupNorm(num_groups=32, num_channels=32)
        # self.norm = nn.BatchNorm2d(num_features=32)
        # self.norm = nn.InstanceNorm2d(num_features=32)

    def forward(self, x, y):
        if y is None:
            y = x
        else:
            y = self.norm(y)

        return y * 2

class G(nn.Module):
    def __init__(self, **ignore_kwargs):
        super().__init__()

        self.mid = Normalize()

    def forward(self, x):

        A = self.mid(x, None)
        B = F.relu(A)

        hs = {}
        hs['A'] = A
        hs['B'] = B
        return hs

class Net(nn.Module):
    def __init__(self):
        super().__init__()

        self.g = G()

        self.norm_1 = Normalize()

    def forward(self, x):
        hs = self.g(x)

        A = hs['A']
        B = hs['B']

        h = self.norm_1(B, A)
        return h

net = Net()
net = net.eval()

x = torch.randn(1, 32, 16, 16)

traced = torch.jit.trace(net, x)
traced.save("test.pt")
```

**Comment by tpoisonooo** [comment]
令人忧愁。

**Comment by Justin62628** [comment]
## pnnx转eval pt的InstanceNorm2d生成的_pnnx.py会漏掉num_features参数
pnnx: pnnx-20230524-windows

- 最小复现
```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class Normalize(nn.Module):
    def __init__(self):
        super().__init__()

        # self.norm = nn.GroupNorm(num_groups=32, num_channels=32)
        # self.norm = nn.BatchNorm2d(num_features=32)
        self.norm = nn.InstanceNorm2d(num_features=32)
        # self.norm = nn.LocalResponseNorm(3)

    def forward(self, x, y):
        if y is None:
            y = x
        else:
            y = self.norm(y)

        y = y * 2

        return y

class G(nn.Module):
    def __init__(self, **ignore_kwargs):
        super().__init__()

        self.norm = Normalize()

    def forward(self, x):

        A = self.norm(x, None)
        B = F.relu(A)

        return A, B

class Net(nn.Module):
    def __init__(self):
        super().__init__()

        self.g = G()

        self.norm_1 = Normalize()

    def forward(self, x):
        hs = self.g(x)

        A, B = hs

        h = self.norm_1(B, A)
        return h

net = Net()
net = net.eval()

x = torch.randn(1, 32, 16, 16)

traced = torch.jit.trace(net, x)
traced.save("test_norm.pt")
```

- 执行`.\pnnx .\test_norm.pt`
- 生成的_pnnx.py:

```python

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()

        self.norm_1_norm = nn.InstanceNorm2d(affine=False, eps=0.000010, track_running_stats=False)

        archive = zipfile.ZipFile('.\test_norm.pnnx.bin', 'r')
        archive.close()
...
    def forward(self, v_0):
        v_1 = (v_0 * 2)
        v_2 = self.norm_1_norm(v_1, v_1, v_1, v_1, v_1)
        v_3 = (v_2 * 2)
        return v_3

def export_torchscript():
    ...
...
```

- 生成的ncnn权重也不可用于推理
![image](https://github.com/Tencent/ncnn/assets/37169106/6935195c-107d-431e-9590-0a4ec610df1a)

- 执行
```python
import numpy as np
import ncnn
import torch

def test_inference():
    torch.manual_seed(0)
    in0 = torch.rand((1,32,24,24), dtype=torch.float32)
    out = []

    with ncnn.Net() as net:
         net.load_param(".\test_norm.ncnn.param")
         net.load_model(".\test_norm.ncnn.bin")

         with net.create_extractor() as ex:
            ex.input("in0", ncnn.Mat(in0.squeeze(0).numpy()).clone())

            _, out0 = ex.extract("out0")
            out.append(torch.from_numpy(np.array(out0)).unsqueeze(0))

    if len(out) == 1:
        return out[0]
    else:
        return tuple(out)

if __name__ == "__main__":
    out = test_inference()
    print(out.shape)
    print(out)
```

输出

```
fopen . est_norm.ncnn.param failed
fopen . est_norm.ncnn.bin failed
find_blob_index_by_name in0 failed
Try
find_blob_index_by_name out0 failed
Try
```

**Comment by Justin62628** [comment]
在torch.jit.trace前面套上torch.jit.optimize_for_inference再过pnnx就能把NormalizeInstance2D处理掉
"""python
    def forward(self, v_0):
        v_1 = (v_0 * 2)
        v_2 = F.instance_norm(input=v_1, bias=None, eps=0.000010, running_mean=None, running_var=None, weight=None)
        v_3 = (v_2 * 2)
        return v_3
"""

**Comment by nihui** [comment]
https://github.com/pytorch/pytorch/pull/102510

**Comment by nihui** [comment]
fixed upstream https://github.com/pytorch/pytorch/commit/7157dfdd4a453829253dc91ab495f4434da227d8

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx
